File: src/greedy/No11.py

from typing import List


class No665:
    def maxArea(self, height: List[int]) -> int:
        num_map = {}
        max_num = 0
        for i in range(len(height)):
            n = height[i]
            max_num = max(max_num, n)

            if n in num_map:
                if len(num_map[n]) < 2:
                    num_map[n].append(i)
                else:
                    num_map[n][1] = i
                
            else:
                num_map[n] = [i]
        
        left = 100000
        right = -1

        res = 0
        for l in range(max_num, 1, -1):
            if l in num_map:
                for k in num_map[l]:
                    left = min(left, k)
                    right = max(right, k)

                res = max(res, l*(right-left))

        return res

print(No665().maxArea([1,1]))

File: src/algorithm/greedy/No11.java
package src.algorithm.greedy;

import java.util.HashMap;

public class No11 {
    int left = Integer.MAX_VALUE;
    int right = Integer.MIN_VALUE;

    HashMap<Integer, Pair> map = new HashMap<>();
    public int maxArea(int[] height) {
        int area = 0;
        int max = Integer.MIN_VALUE;
        for (int i = 0; i < height.length; i++) {
            Pair pair = map.getOrDefault(height[i], new Pair(i, i));
            pair.l = Math.min(pair.l, i);
            pair.r = Math.max(pair.r, i);
            map.put(height[i], pair);
            max = Math.max(max, height[i]);
        }

        for (int i = max; i >= 0; i--) {
            Pair pair = map.get(i);
            if (pair != null) {
                left = Math.min(left, pair.l);
                right = Math.max(right, pair.r);
                area = Math.max(area, i * (right - left));
            }
        }
        return area;
    }

    private static class Pair {
        int l;
        int r;

        public Pair(int l, int r) {
            this.l = l;
            this.r = r;
        }
    }

    public static void main(String[] args) {
        int[] height = {1,8,6,2,5,4,8,3,7};
        new No11().maxArea(height);
    }
}
